Fix flaky test `TestProxySSHJumpHost`

Joerger opened the pull request:
Attempt to fix https://github.com/gravitational/teleport/issues/39578

I turned of SSH connection resumption as it seems like that was causing the flake.

I also switch to `testserver.MakeTestServer` because `newTestSuite` was having issues in parallel runs, likely because the file descriptors for listeners are handled differently. This also improved the test speed from 9 to 8 seconds, so maybe it will have an impact.

The test was also outdated as it was testing if we issued a valid leaf certificate for `tsh proxy ssh -J`. This functionality was removed in favor of using `load_all_cas` since the leaf trusts the root cert and we just need the leaf CA.

Joerger committed:
Use testServer.MakeTestServer to fix issue with parallel runs using the same ports.

zmb3 commented on tool/teleport/testenv/test_server.go:
```diff
@@ -330,6 +330,36 @@ func WithAuthPreference(authPref types.AuthPreference) TestServerOptFunc {
 	})
 }
 
+func SetupTrustedCluster(ctx context.Context, t *testing.T, rootServer, leafServer *service.TeleportProcess, additionalRoleMappings ...types.RoleMapping) {
```
Are there other trusted cluster tests that should use this in the future?

I'm concerned about leaving multiple ways to do the same thing in the codebase and having people reintroduce flaky tests in the future by picking the wrong one.

I'm not suggesting any larger changes in this PR, but maybe a comment or TODO would help.

Joerger replied:
I don't think this particular change contributes to the flakiness here, but agreed. I added deprecation comments. Ideally we would only use `tool/teleport/testenv` for test servers so that any fixes apply to all tests.

Joerger committed:
Add deprecation comments for test server helpers outside of tools/teleport/testenv.

zmb3 reviewed (approved):
Looks good, assuming the `// deprecated` comments don't cause lint failures.

Joerger committed:
Update tool/teleport/testenv/test_server.go

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

Joerger committed:
Disable SSH resumption

Enable LoadAllCAs

Update comments

rosstimothy committed:
Prevent context.Canceled from being wrapped across transport stream

trail doesn't properly handle conversions to and from errors that
are context.Canceled which results in code paths trying to evaluate
errors.Is(err, context.Canceled) failing. The error recevied is
an interceptors.RemoteError, which wraps a trace.TraceErr, which
contains a status.Status with codes.Canceled. To work around this
until trail is updated the server half of the stream was updated
to return an io.EOF if if encouters a canceled error.

zmb3 commented:
/excludeflake *
